src/node/audio_buffer_source.rs

```diff
@@ -436,6 +432,113 @@ impl AudioProcessor for AudioBufferSourceRenderer {
+                self.render_state.started = true;
+            }
+
+            // check if buffer ends within this block
+            if self.render_state.buffer_time + block_duration > buffer_duration
+                || self.render_state.buffer_time + block_duration > duration
+                || current_time + block_duration > stop_time
+            {
+                let buffer_time = self.render_state.buffer_time;
+                let mut apply_offset = 0;
+
+                buffer
+                    .channels()
+                    .iter()
+                    .zip(output.channels_mut().iter_mut())
+                    .for_each(|(buffer_channel, output_channel)| {
+                        // we need to recompute that for each channel
+                        let buffer_channel = buffer_channel.as_slice();
+                        let mut start_index = (buffer_time * sample_rate).round() as usize;
+                        let end_index = if current_time + block_duration > stop_time
+                            || self.render_state.buffer_time + block_duration > duration
+                        {
+                            let dt = (stop_time - current_time)
+                                .min(duration - self.render_state.buffer_time);
+                            let end_buffer_time = self.render_state.buffer_time + dt;
+                            (end_buffer_time * sample_rate).round() as usize
+                        } else {
+                            buffer.length()
+                        };
+                        let mut offset = 0;
+
+                        for (index, o) in output_channel.iter_mut().enumerate() {
+                            let mut buffer_index = start_index + index - offset;
+
+                            *o = if buffer_index < end_index {
+                                buffer_channel[buffer_index]
+                            } else {
+                                if loop_ && buffer_index == end_index {
+                                    start_index = 0;
+                                    offset = index;
+                                    buffer_index = 0;
+                                    // extract value to compute buffer_time
+                                    apply_offset = offset;
+                                }
+
+                                if loop_ {
+                                    buffer_channel[buffer_index]
+                                } else {
+                                    0.
+                                }
+                            };
+                        }
+                    });
+
+                if apply_offset != 0 {
+                    self.render_state.buffer_time = ((RENDER_QUANTUM_SIZE - apply_offset) as f64
+                        / sample_rate)
+                        % buffer_duration;
+                } else {
+                    self.render_state.buffer_time += block_duration;
+                }
+            } else {
```

Comment:
With looping and offsets not part of the fast path, I think we only need this following snippet (memcopy).
Of course you still need to pad with zeroes if the end_index < 128 + start_index